File: packages/js/core/src/models/mamba.ts
import { Tensor } from '../ir/tensor.js';
import { ConvND, Gemm, RMSNorm, Silu } from '../primitives.js';

function getParam(name: string, shape: number[], dtype: any = 'float32'): Tensor {
  return new Tensor(name, shape, dtype, false, false, new Float32Array());
}

function recordOp(opType: string, inputs: Tensor[], attr?: any): Tensor {
  const dtype = inputs[0]?.dtype ?? 'float32';
  return new Tensor(`${opType}_out`, [], dtype, false, false, new Float32Array());
}

export class StateSpace {
  public dModel: number;
  public dState: number;
  public dConv: number;
  public expand: number;

  constructor(dModel: number, dState: number, dConv: number, expand: number = 2) {
    this.dModel = dModel;
    this.dState = dState;
    this.dConv = dConv;
    this.expand = expand;
  }

  call(x: Tensor, dt: Tensor, a: Tensor, b: Tensor, c: Tensor, d: Tensor): Tensor {
    return recordOp('StateSpace', [x, dt, a, b, c, d], {
      d_model: this.dModel,
      d_state: this.dState,
      d_conv: this.dConv,
      expand: this.expand,
    });
  }
}

export class MambaBlock {
  public prefix: string;
  public dModel: number;
  public dState: number;
  public dConv: number;
  public expand: number;

  public norm: RMSNorm;
  public inProj: Gemm;
  public conv1d: ConvND;
  public act: Silu;
  public xProj: Gemm;
  public dtProj: Gemm;
  public ssm: StateSpace;
  public outProj: Gemm;

  constructor(dModel: number, dState: number, dConv: number, expand: number, prefix: string = '') {
    this.prefix = prefix;
    this.dModel = dModel;
    this.dState = dState;
    this.dConv = dConv;
    this.expand = expand;

    this.norm = new RMSNorm([dModel]);
    this.inProj = new Gemm(1.0, 1.0, 0, 1);

    const dInner = dModel * expand;
    this.conv1d = new ConvND(1, dInner, dInner, dConv, 1, dConv - 1, 1, dInner, false);
    this.act = new Silu();

    this.xProj = new Gemm(1.0, 1.0, 0, 1);
    this.dtProj = new Gemm(1.0, 1.0, 0, 1);

    this.ssm = new StateSpace(dModel, dState, dConv, expand);

    this.outProj = new Gemm(1.0, 1.0, 0, 1);
  }

  call(x: Tensor): Tensor {
    const identity = x;
    const xNorm = this.norm.call(x, getParam(`${this.prefix}.norm.weight`, [this.dModel]));

    const dInner = this.dModel * this.expand;
    const xz = this.inProj.call(
      xNorm,
      getParam(`${this.prefix}.in_proj.weight`, [dInner * 2, this.dModel]),
    );

    const xInner = recordOp('Slice', [
      xz,
      recordOp('Constant', [], { value: [0], dtype: 7 }),
      recordOp('Constant', [], { value: [dInner], dtype: 7 }),
      recordOp('Constant', [], { value: [2], dtype: 7 }),
    ]);
    const z = recordOp('Slice', [
      xz,
      recordOp('Constant', [], { value: [dInner], dtype: 7 }),
      recordOp('Constant', [], { value: [dInner * 2], dtype: 7 }),
      recordOp('Constant', [], { value: [2], dtype: 7 }),
    ]);

    let xInnerT = recordOp('Transpose', [xInner], { perm: [0, 2, 1] });
    let xInnerConv = this.conv1d.call(
      xInnerT,
      getParam(`${this.prefix}.conv1d.weight`, [dInner, 1, this.dConv]),
      getParam(`${this.prefix}.conv1d.bias`, [dInner]),
    );
    xInnerConv = recordOp('Transpose', [xInnerConv], { perm: [0, 2, 1] });

    const seqLen = recordOp(
      'Gather',
      [recordOp('Shape', [x]), recordOp('Constant', [], { value: [1], dtype: 7 })],
      { axis: 0 },
    );
    xInnerConv = recordOp('Slice', [
      xInnerConv,
      recordOp('Constant', [], { value: [0], dtype: 7 }),
      seqLen,
      recordOp('Constant', [], { value: [1], dtype: 7 }),
    ]);

    const xAct = this.act.call(xInnerConv);

    const xDtBC = this.xProj.call(
      xAct,
      getParam(`${this.prefix}.x_proj.weight`, [this.dState * 2 + 1, dInner]),
    );

    let dt = recordOp('Slice', [
      xDtBC,
      recordOp('Constant', [], { value: [0], dtype: 7 }),
      recordOp('Constant', [], { value: [1], dtype: 7 }),
      recordOp('Constant', [], { value: [2], dtype: 7 }),
    ]);
    const b = recordOp('Slice', [
      xDtBC,
      recordOp('Constant', [], { value: [1], dtype: 7 }),
      recordOp('Constant', [], { value: [this.dState + 1], dtype: 7 }),
      recordOp('Constant', [], { value: [2], dtype: 7 }),
    ]);
    const c = recordOp('Slice', [
      xDtBC,
      recordOp('Constant', [], { value: [this.dState + 1], dtype: 7 }),
      recordOp('Constant', [], { value: [this.dState * 2 + 1], dtype: 7 }),
      recordOp('Constant', [], { value: [2], dtype: 7 }),
    ]);

    dt = this.dtProj.call(
      dt,
      getParam(`${this.prefix}.dt_proj.weight`, [dInner, 1]),
      getParam(`${this.prefix}.dt_proj.bias`, [dInner]),
    );

    const a = getParam(`${this.prefix}.A_log`, [dInner, this.dState]);
    const d = getParam(`${this.prefix}.D`, [dInner]);

    let y = this.ssm.call(xAct, dt, a, b, c, d);

    const zAct = this.act.call(z);
    y = recordOp('Mul', [y, zAct]);

    const out = this.outProj.call(
      y,
      getParam(`${this.prefix}.out_proj.weight`, [this.dModel, dInner]),
    );
    return recordOp('Add', [identity, out]);
  }
}

export class Mamba {
  public vocabSize: number;
  public dModel: number;
  public blocks: MambaBlock[];
  public norm: RMSNorm;
  public lmHead: Gemm;

  constructor(
    vocabSize: number = 50277,
    dModel: number = 768,
    nLayer: number = 24,
    dState: number = 16,
    dConv: number = 4,
    expand: number = 2,
  ) {
    this.vocabSize = vocabSize;
    this.dModel = dModel;

    this.blocks = [];
    for (let i = 0; i < nLayer; i++) {
      this.blocks.push(new MambaBlock(dModel, dState, dConv, expand, `blocks.${i}`));
    }

    this.norm = new RMSNorm([dModel]);
    this.lmHead = new Gemm(1.0, 1.0, 0, 1);
  }

  call(inputIds: Tensor): Tensor {
    let x = recordOp(
      'Gather',
      [getParam('embedding.weight', [this.vocabSize, this.dModel]), inputIds],
      { axis: 0 },
    );

    for (const block of this.blocks) {
      x = block.call(x);
    }

    x = this.norm.call(x, getParam('norm.weight', [this.dModel]));
    x = this.lmHead.call(x, getParam('lm_head.weight', [this.vocabSize, this.dModel]));
    return x;
  }
}

export function mamba130m(): Mamba {
  return new Mamba(50277, 768, 24, 16, 4, 2);
}

File: packages/python/onnx9000-core/src/onnx9000/core/models/mamba.py
"""Module docstring."""

from typing import Any

from onnx9000.core.ir import Tensor, Variable
from onnx9000.core.ops import add, mul
from onnx9000.core.primitives import ConvND, Gemm, RMSNorm, Silu, StateSpace


def get_param(name: str, shape: list[int], dtype: int = 1) -> Variable:  # noqa: D103
    return Variable(name=name, shape=shape, dtype=dtype)


class MambaBlock:  # noqa: D101
    def __init__(self, d_model: int, d_state: int, d_conv: int, expand: int, prefix: str = ""):  # noqa: D107
        self.prefix = prefix
        self.d_model = d_model
        self.d_state = d_state
        self.d_conv = d_conv
        self.expand = expand

        self.norm = RMSNorm((d_model,))
        self.in_proj = Gemm(trans_b=1)

        self.conv1d = ConvND(
            1,
            d_model * expand,
            d_model * expand,
            kernel_size=d_conv,
            groups=d_model * expand,
            padding=d_conv - 1,
        )
        self.act = Silu()

        self.x_proj = Gemm(trans_b=1)
        self.dt_proj = Gemm(trans_b=1)

        self.ssm = StateSpace(d_model, d_state, d_conv, expand)

        self.out_proj = Gemm(trans_b=1)

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        identity = x
        x_norm = self.norm(x, get_param(f"{self.prefix}.norm.weight", [self.d_model]))

        # in_proj outputs [B, L, 2 * d_inner]
        xz = self.in_proj(
            x_norm,
            get_param(
                f"{self.prefix}.in_proj.weight", [self.d_model * self.expand * 2, self.d_model]
            ),
        )

        # Split into x and z
        from onnx9000.core.ops import constant
        from onnx9000.core.ops import slice as slice_op

        d_inner = self.d_model * self.expand
        x_inner = slice_op(
            xz, constant([0], dtype=7), constant([d_inner], dtype=7), constant([2], dtype=7)
        )
        z = slice_op(
            xz,
            constant([d_inner], dtype=7),
            constant([d_inner * 2], dtype=7),
            constant([2], dtype=7),
        )

        # Conv1D needs [B, C, L]
        from onnx9000.core.ops import transpose

        x_inner_t = transpose(x_inner, perm=[0, 2, 1])
        x_inner_conv = self.conv1d(
            x_inner_t,
            get_param(f"{self.prefix}.conv1d.weight", [d_inner, 1, self.d_conv]),
            get_param(f"{self.prefix}.conv1d.bias", [d_inner]),
        )
        x_inner_conv = transpose(x_inner_conv, perm=[0, 2, 1])

        # Slice to original L since causal padding adds d_conv - 1
        from onnx9000.core.ops import gather, shape

        seq_len = gather(shape(x), constant([1], dtype=7), axis=0)
        x_inner_conv = slice_op(
            x_inner_conv, constant([0], dtype=7), seq_len, constant([1], dtype=7)
        )

        x_act = self.act(x_inner_conv)

        # SSM parameters
        x_dt_B_C = self.x_proj(
            x_act, get_param(f"{self.prefix}.x_proj.weight", [self.d_state * 2 + 1, d_inner])
        )
        dt = slice_op(
            x_dt_B_C, constant([0], dtype=7), constant([1], dtype=7), constant([2], dtype=7)
        )
        B = slice_op(
            x_dt_B_C,
            constant([1], dtype=7),
            constant([self.d_state + 1], dtype=7),
            constant([2], dtype=7),
        )
        C = slice_op(
            x_dt_B_C,
            constant([self.d_state + 1], dtype=7),
            constant([self.d_state * 2 + 1], dtype=7),
            constant([2], dtype=7),
        )

        dt = self.dt_proj(
            dt,
            get_param(f"{self.prefix}.dt_proj.weight", [d_inner, 1]),
            get_param(f"{self.prefix}.dt_proj.bias", [d_inner]),
        )

        A = get_param(f"{self.prefix}.A_log", [d_inner, self.d_state])
        D = get_param(f"{self.prefix}.D", [d_inner])

        # State Space Op
        y = self.ssm(x_act, dt, A, B, C, D)

        # Gate
        z_act = self.act(z)
        y = mul(y, z_act)

        # Out
        out = self.out_proj(y, get_param(f"{self.prefix}.out_proj.weight", [self.d_model, d_inner]))
        return add(identity, out)


class Mamba:  # noqa: D101
    def __init__(  # noqa: D107
        self,
        vocab_size: int = 50277,
        d_model: int = 768,
        n_layer: int = 24,
        d_state: int = 16,
        d_conv: int = 4,
        expand: int = 2,
    ):
        self.vocab_size = vocab_size
        self.d_model = d_model

        self.blocks = [
            MambaBlock(d_model, d_state, d_conv, expand, prefix=f"blocks.{i}")
            for i in range(n_layer)
        ]
        self.norm = RMSNorm((d_model,))
        self.lm_head = Gemm(trans_b=1)

    def __call__(self, input_ids: Tensor) -> Tensor:  # noqa: D102
        from onnx9000.core.ops import gather

        x = gather(
            get_param("embedding.weight", [self.vocab_size, self.d_model]), input_ids, axis=0
        )

        for block in self.blocks:
            x = block(x)

        x = self.norm(x, get_param("norm.weight", [self.d_model]))
        x = self.lm_head(x, get_param("lm_head.weight", [self.vocab_size, self.d_model]))
        return x


def mamba_130m(**kwargs: Any) -> Mamba:  # noqa: D103
    return Mamba(
        vocab_size=50277, d_model=768, n_layer=24, d_state=16, d_conv=4, expand=2, **kwargs
    )
